dcos/auth.py

```diff
@@ -0,0 +1,308 @@
+    else:
+        return
+
+
+def _get_dcostoken_by_oidc_implicit_flow(dcos_url):
+    """
+    Get DC/OS Authentication Token by OIDC implicit flow
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :returns: DC/OS Authentication Token
+    :rtype: str
+    """
+
+    oauth_login = 'login?redirect_uri=urn:ietf:wg:oauth:2.0:oob'
+    url = urllib.parse.urljoin(dcos_url, oauth_login)
+    oidc_token = _prompt_user_for_token(url, "OpenID Connect ID")
+    creds = {"token": oidc_token}
+    return _get_dcostoken_by_post_with_creds(dcos_url, creds)
+
+
+def servicecred_auth(dcos_url, username, key_path):
+    """
+    Get DC/OS Authentication token by browser prompt
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :param username: username user for authentication
+    :type username: str
+    :param key_path: path to service key
+    :param key_path: str
+    :rtype: None
+    """
+
+    creds = {
+        'uid': username,
+        'token': jwt.encode(
+            {
+                'exp': int(time.time()+5*60),
+                'uid': username
+            },
+            util.read_file(key_path),
+            algorithm='RS256')
+        .decode('ascii')
+    }
+
+    dcos_token = _get_dcostoken_by_post_with_creds(dcos_url, creds)
+    if not dcos_token:
+        raise DCOSException("Authentication failed")
+    else:
+        return
+
+
+def browser_prompt_auth(dcos_url, provider_info):
+    """
+    Get DC/OS Authentication token by browser prompt
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :param provider_info: info about provider to auth with
+    :param provider_info: str
+    :rtype: None
+    """
+
+    start_flow_url = provider_info["client"]["start_flow_url"]
+    if not urlparse(start_flow_url).netloc:
+        start_flow_url = urljoin(dcos_url, start_flow_url)
+
+    dcos_token = _prompt_user_for_token(start_flow_url, "DC/OS Authentication")
+
+    # verify token
+    endpoint = '/pkgpanda/active.buildinfo.full.json'
+    url = urllib.parse.urljoin(dcos_url, endpoint)
+    response = http._request('HEAD', url, auth=http.DCOSAcsToken(dcos_token))
+    if response.status_code in [200, 403]:
+        config.set_val("core.dcos_acs_token", dcos_token)
+    else:
+        raise DCOSException("Authentication failed")
+
+
+def header_challenge_auth(dcos_url):
+    """
+    Triggers authentication using scheme specified in www-authenticate header.
+
+    Raises exception if authentication fails.
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :rtype: None
+    """
+
+    # hit protected endpoint which will prompt for auth if cluster has auth
+    endpoint = '/pkgpanda/active.buildinfo.full.json'
+    url = urllib.parse.urljoin(dcos_url, endpoint)
+    response = http._request('HEAD', url)
+    auth_scheme = _get_auth_scheme(response)
+
+    i = 0
+    while i < 3 and response.status_code == 401:
+        if auth_scheme == "oauthjwt":
```

Comment:
I believe that for the case distinctions (as in `if auth_scheme == "oauthjwt":`) we should add some comments about what we believe to have learned about the cluster from here.